### PR TITLE
ARROW-1611 Do not read bitmap when length is zero, add internal::Bitm…

### PR DESCRIPTION
**Pull request by renesugar**
Do not read or write bitmap when length is zero, add internal::BitMapWriter

### CODE REVIEW

**Comment by wesm** [comment]
Note for future reference it's not necessary to open a new PR, you can edit the title by hitting the "Edit" button in the top right. 

This has some compiler warnings:

```
In file included from /home/travis/build/apache/arrow/cpp/src/arrow/buffer.h:29:0,
                 from /home/travis/build/apache/arrow/cpp/src/arrow/buffer.cc:18:
/home/travis/build/apache/arrow/cpp/src/arrow/util/bit-util.h: In member function ‘void arrow::internal::BitmapWriter::Set()’:
/home/travis/build/apache/arrow/cpp/src/arrow/util/bit-util.h:108:30: error: conversion to ‘uint8_t {aka unsigned char}’ from ‘int’ may alter its value [-Werror=conversion]
   void Set() { current_byte_ |= (1 << bit_offset_); }
                              ^
/home/travis/build/apache/arrow/cpp/src/arrow/util/bit-util.h: In member function ‘void arrow::internal::BitmapWriter::Clear()’:
/home/travis/build/apache/arrow/cpp/src/arrow/util/bit-util.h:110:32: error: conversion to ‘uint8_t {aka unsigned char}’ from ‘int’ may alter its value [-Werror=conversion]
   void Clear() { current_byte_ &= ~(1 << bit_offset_); }
                                ^
cc1plus: all warnings being treated as errors
[28/137] Building CXX object CMakeFiles/arrow_objlib.dir/src/arrow/array.cc.o
FAILED: CMakeFiles/arrow_objlib.dir/src/arrow/array.cc.o 
/usr/bin/ccache /usr/lib/ccache/g++-4.9  -DARROW_EXTRA_ERROR_CONTEXT -DARROW_NO_DEPRECATED_API -DARROW_WITH_BROTLI -DARROW_WITH_LZ4 -DARROW_WITH_SNAPPY -DARROW_WITH_ZLIB -DARROW_WITH_ZSTD -isystem /home/travis/build/apache/arrow/cpp-toolchain/include -isystem googletest_ep-prefix/src/googletest_ep/include -isystem gbenchmark_ep/src/gbenchmark_ep-install/include -isystem /home/travis/build/apache/arrow/cpp/thirdparty/hadoop/include -I/home/travis/build/apache/arrow/cpp/src -ggdb -O0 -Wall -std=c++11 -msse3 -Wconversion -Wno-sign-conversion -Werror -g -fPIC   -std=gnu++11 -MD -MT CMakeFiles/arrow_objlib.dir/src/arrow/array.cc.o -MF CMakeFiles/arrow_objlib.dir/src/arrow/array.cc.o.d -o CMakeFiles/arrow_objlib.dir/src/arrow/array.cc.o -c /home/travis/build/apache/arrow/cpp/src/arrow/array.cc
In file included from /home/travis/build/apache/arrow/cpp/src/arrow/buffer.h:29:0,
                 from /home/travis/build/apache/arrow/cpp/src/arrow/array.h:27,
                 from /home/travis/build/apache/arrow/cpp/src/arrow/array.cc:18:
/home/travis/build/apache/arrow/cpp/src/arrow/util/bit-util.h: In member function ‘void arrow::internal::BitmapWriter::Set()’:
/home/travis/build/apache/arrow/cpp/src/arrow/util/bit-util.h:108:30: error: conversion to ‘uint8_t {aka unsigned char}’ from ‘int’ may alter its value [-Werror=conversion]
   void Set() { current_byte_ |= (1 << bit_offset_); }
                              ^
/home/travis/build/apache/arrow/cpp/src/arrow/util/bit-util.h: In member function ‘void arrow::internal::BitmapWriter::Clear()’:
/home/travis/build/apache/arrow/cpp/src/arrow/util/bit-util.h:110:32: error: conversion to ‘uint8_t {aka unsigned char}’ from ‘int’ may alter its value [-Werror=conversion]
   void Clear() { current_byte_ &= ~(1 << bit_offset_); }
                                ^
cc1plus: all warnings being treated as errors
```

You might want to use the appropriate bitmasks from https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/bit-util.h#L152

File: cpp/src/arrow/util/bit-util-test.cc [hunk 1]
```diff
@@ -93,6 +93,39 @@ TEST(BitmapReader, DoesNotReadOutOfBounds) {
   }
 }
 
+TEST(BitmapWriter, DoesNotWriteOutOfBounds) {
```
Comment (wesm) [inline cpp/src/arrow/util/bit-util-test.cc]: Can you also add a test for `BitmapReader` where length is 0?

**Review by wesm — COMMENTED** [review]
Thanks for adding a test, minor comment, and then there's just getting the build to pass

**Comment by wesm** [comment]
Feel free also to make me a collaborator on your fork

**Comment by wesm** [comment]
I addressed these points in https://github.com/apache/arrow/pull/1137 to unblock the 0.7.1 release candidate. I will merge once the build is green there

I may wait a few hours for https://github.com/apache/arrow/pull/1137 to possibly get done before cutting the RC

**Comment by renesugar** [comment]
Updated pull request branch. I was tracking down the cause of the build failure; I'll open a separate issue for what I found.

**Pull request closed without merging**